pkg/engine/constraints.go

```diff
@@ -0,0 +1,183 @@
+		Operator ConstraintOperator `yaml:"operator"`
+		Node     core.ResourceId    `yaml:"node"`
+		Edge     Edge               `yaml:"edge"`
+	}
+
+	// ConstructConstraint is a struct that represents constraints that can be applied on a specific construct in the resource graph
+	ConstructConstraint struct {
+		Operator   ConstraintOperator `yaml:"operator"`
+		Target     core.ResourceId    `yaml:"target"`
+		Type       string             `yaml:"type"`
+		Attributes map[string]any     `yaml:"attributes"`
+	}
+
+	// EdgeConstraint is a struct that represents constraints that can be applied on a specific edge in the resource graph
+	EdgeConstraint struct {
+		Operator ConstraintOperator `yaml:"operator"`
+		Target   Edge               `yaml:"target"`
+		Node     core.ResourceId    `yaml:"node"`
+	}
+
+	// NodeConstraint is a struct that represents constraints that can be applied on a specific node in the resource graph.
+	// NodeConstraints are used to control intrinsic properties of a node in the resource graph
+	NodeConstraint struct {
+		BaseConstraint
+		Target   core.ResourceId
+		Property string
+		Value    any
+	}
+
+	// ConstraintScope is an enum that represents the different scopes that a constraint can be applied to
+	ConstraintScope string
+	// ConstraintOperator is an enum that represents the different operators that can be applied to a constraint
+	ConstraintOperator string
+)
+
+const (
+	ApplicationConstraintScope ConstraintScope = "application"
+	ConstructConstraintScope   ConstraintScope = "construct"
+	EdgeConstraintScope        ConstraintScope = "edge"
+	NodeConstraintScope        ConstraintScope = "node"
+
+	MustExistConstraintOperator      ConstraintOperator = "must_exist"
+	MustContainConstraintOperator    ConstraintOperator = "must_contain"
+	MustNotExistConstraintOperator   ConstraintOperator = "must_not_exist"
+	MustNotContainConstraintOperator ConstraintOperator = "must_not_contain"
+	AddConstraintOperator            ConstraintOperator = "add"
+	EqualsConstraintOperator         ConstraintOperator = "equals"
+)
+
+func (b *ApplicationConstraint) Scope() ConstraintScope {
+	return ApplicationConstraintScope
+}
+
+func (b *ConstructConstraint) Scope() ConstraintScope {
+	return ConstructConstraintScope
+}
+
+func (b *EdgeConstraint) Scope() ConstraintScope {
+	return EdgeConstraintScope
+}
+
+// DecodeYAMLNode is a helper function that decodes a yaml node into a struct representing different constraints
+func DecodeYAMLNode[T Constraint](node *yaml.Node) (constraint T, err error) {
+	constraint = reflect.New(reflect.TypeOf(constraint).Elem()).Interface().(T)
+	err = extraFields(node, reflect.ValueOf(constraint))
+	if err != nil {
+		return constraint, err
+	}
+	err = node.Decode(constraint)
+	return constraint, err
+}
+
+// ParseConstraintsFromFile is a helper function that parses a yaml file into a map of constraints
+func ParseConstraintsFromFile(path string) (map[ConstraintScope][]Constraint, error) {
```

Comment:
I'm not sure this is the structure we want. We may want to allow grouping and ordering from the caller side so that the constraints are evaluated in a specific, predictable order.

Reply:
lets chat about this because this is likely something we want to define. 

My thought is that we have levels to constraints and it must go in order of 
1. application
2. Construct
3. edge
4. node

if were saying we want even more ordering within that we can build it in, but are we saying well hold the order of the yaml file?